articles/qos_configurability.md

```diff
@@ -0,0 +1,399 @@
+
+## Design goals and other constraints
+
+This proposal was written with the following goals in mind:
+
+- The node's author should be able to control which entities have reconfigurable QoS, and if a profile is valid or not.
+  - Rationale: Allowing to change all QoS policies could break a contract the node's implementer assumed.
+  For example, the node's implementer may have assumed that the data was durable and not volatile.
+- QoS should not be dynamically reconfigurable.
+  - Rationale: Allowing runtime reconfiguration of QoS is complex, and in many cases it involves tearing down the previous entity and creating a new one.
+  That process can cause loss of messages, and thus providing an automatic mechanism for it is not ideal.
+- It should be possible to reuse the same QoS profile for different entities.
+  - Rationale: Not having a way of setting the same QoS profile to different entities can cause errors, considering that in many cases it's desired to have matching qos.
+
+The following constraint was also considered in the design:
+
+- Only apply to publisher/subscriber QoS.
+  - Rationale: QoS profiles for services and actions are not well defined currently, and modifying them is not recommended.
+  Thus, making them reconfigurable doesn't make much sense right now.
+
+## Adding a new mechanism vs using parameters
+
+A previous proposal (see PR [#296](https://github.com/ros2/design/pull/296)) suggested using a new file format for reconfiguring QoS policies.
+Some concerns were raised by the community:
+1. It adds a different mechanism to the ones that already exist.
+    Newcomers have to learn yet another thing.
+1. QoS settings are not conceptually different to other parameters, e.g.: publishing rate.
+1. It adds a new mechanism to configure QoS profiles.
+
+Based on that feedback, this PR will explore a design based on parameters.
+
+## Introduction
+
+Current code will not allow QoS reconfigurability if not changed:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast{10});
+```
+
+To make reconfigurability easy, only a flag is needed to automatically create the parameters:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast{10},
+  QosOverridingOptions{true});  // allow_reconfigurable all qos
+```
+
+That will automatically declare the parameters for reconfiguring the QoS policies that can be overridden in the following way:
+
+```yaml
+/my/ns/node_name:
+  ros__parameters:
+    qos_profiles:  # group for all the qos settings
+      publisher:
+        'my/fully/qualified/topic/name/here': 
+          reliability: reliable
+          history_depth: 100
+          history: keep_last
+```
+
+A callback to accept/reject the configured profile can also be specified:
+
+```cpp
+node->create_publisher(
+  "chatter",
+  KeepLast(10),
+  [] (const QoSProfile & qos) -> bool {
+      return qos.is_reliable();  // only accept reliable communication, we should add getters to rclcpp::QoSProfile class for this.
+  });
```

Comment:
This is fine for brevity of the example, but again I think this should be in the `PublisherOptions`.

Reply:
Clarified in https://github.com/ros2/design/pull/300/commits/c5d88491678c9cba5073e8cbd9910d1e56d0ddbc.